src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,395 @@
+  RAY_CHECK(actor && worker);
+  actor->UpdateAddress(worker->GetAddress());
+  RAY_LOG(INFO) << "Start creating actor " << actor->GetActorID() << " on worker "
+                << worker->GetWorkerID() << " at node " << actor->GetNodeID();
+  worker->CreateActor(actor->GetCreationTaskSpecification(), [this, actor, worker] {
+    // This callback is invoked only when the actor is created successfully.
+    RAY_LOG(INFO) << "Succeeded in creating actor " << actor->GetActorID()
+                  << " on worker " << worker->GetWorkerID() << " at node "
+                  << actor->GetNodeID();
+    RAY_CHECK(actor->GetActorID() ==
+              RemoveWorkerInPhaseOfCreating(worker->GetNodeID(), worker->GetWorkerID()));
+    schedule_success_handler_(actor);
+  });
+}
+
+std::shared_ptr<rpc::GcsNodeInfo> GcsActorScheduler::SelectNodeRandomly() const {
+  auto &alive_nodes = gcs_node_manager_.GetAllAliveNodes();
+  if (alive_nodes.empty()) {
+    return nullptr;
+  }
+
+  static std::mt19937_64 gen_(
+      std::chrono::high_resolution_clock::now().time_since_epoch().count());
+  std::uniform_int_distribution<int> distribution(0, alive_nodes.size() - 1);
+  int key_index = distribution(gen_);
+  int index = 0;
+  auto iter = alive_nodes.begin();
+  for (; index != key_index && iter != alive_nodes.end(); ++index, ++iter)
+    ;
+  return iter->second;
+}
+
+std::shared_ptr<WorkerLeaseInterface> GcsActorScheduler::GetOrConnectLeaseClient(
+    const rpc::Address &raylet_address) {
+  auto node_id = ClientID::FromBinary(raylet_address.raylet_id());
+  auto iter = remote_lease_clients_.find(node_id);
+  if (iter == remote_lease_clients_.end()) {
+    auto lease_client = lease_client_factory_(raylet_address);
+    iter = remote_lease_clients_.emplace(node_id, std::move(lease_client)).first;
+  }
+  return iter->second;
+}
+
+void GcsActorScheduler::AddActorInPhaseOfLeasing(std::shared_ptr<GcsActor> actor) {
+  node_to_actors_when_leasing_[actor->GetNodeID()].emplace(actor->GetActorID());
+}
+
+void GcsActorScheduler::RemoveActorInPhaseOfLeasing(std::shared_ptr<GcsActor> actor) {
+  auto iter = node_to_actors_when_leasing_.find(actor->GetNodeID());
+  if (iter != node_to_actors_when_leasing_.end()) {
+    iter->second.erase(actor->GetActorID());
+    if (iter->second.empty()) {
+      node_to_actors_when_leasing_.erase(iter);
+    }
+  }
+}
+
+void GcsActorScheduler::AddWorkerInPhaseOfCreating(
+    std::shared_ptr<GcsLeasedWorker> leased_worker) {
+  RAY_CHECK(leased_worker != nullptr);
+  auto node_id = leased_worker->GetNodeID();
+  node_to_workers_when_creating_[node_id].emplace(leased_worker->GetWorkerID(),
+                                                  leased_worker);
+}
+
+ActorID GcsActorScheduler::RemoveWorkerInPhaseOfCreating(const ClientID &node_id,
+                                                         const WorkerID &worker_id) {
+  ActorID assigned_actor_id;
```

Comment:
> Assert that the deletion was successful.

There is not necessarily an actor being scheduled on the specified worker. If there is one, the `assigned_actor_id` inside the leased worker will be returned, otherwise, nil will be returned. I have checked the return value in the callback of `worker->CreateActor`